src/wp-includes/blocks.php

```diff
@@ -760,6 +763,86 @@ function get_hooked_blocks( $name ) {
 	return $hooked_blocks;
 }
 
+/**
+ * Returns a function that...
+ *
+ * @since 6.4.0
+ * @access private
+ *
+ * @param WP_Block_Template|array $context A block template, template part, or pattern that the blocks belong to.
+ * @return callable A function that returns the serialized markup for the given block.
+ */
+function make_before_block_visitor( $context ) {
+	return function( &$block, $parent = null, $prev = null ) use ( $context ) {
+		_inject_theme_attribute_in_template_part_block( $block );
+
+		$markup = '';
+
+		if ( $parent && ! $prev ) {
+			// Candidate for first-child insertion.
+			$hooked_blocks_for_parent = get_hooked_blocks( $parent['blockName'] );
+			foreach ( $hooked_blocks_for_parent as $hooked_block_type => $relative_position ) {
+				if ( 'first_child' === $relative_position ) {
+					$hooked_block_markup = get_comment_delimited_block_content( $hooked_block_type, array(), '' );
+					/** This filter is documented in wp-includes/blocks.php */
+					$markup .= apply_filters( 'insert_hooked_block', $hooked_block_markup, $hooked_block_type, $relative_position, $parent, $context );
+				}
+			}
+		}
+
+		$hooked_blocks = get_hooked_blocks( $block['blockName'] );
+		foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
+			if ( 'before' === $relative_position ) {
+				$hooked_block_markup = get_comment_delimited_block_content( $hooked_block_type, array(), '' );
+				/**
+				 * Filters the serialized markup of a hooked block.
+				 *
+				 * @since 6.4.0
+				 *
+				 * @param string                  $hooked_block_markup The serialized markup of the hooked block.
+				 * @param string                  $hooked_block_type   The type of the hooked block.
+				 * @param string                  $relative_position   The relative position of the hooked block.
+				 *                                                     Can be one of 'before', 'after', 'first_child', or 'last_child'.
+				 * @param array                   $block               The anchor block.
+				 * @param WP_Block_Template|array $context             The block template, template part, or pattern that the anchor block belongs to.
+				 */
+				$markup .= apply_filters( 'insert_hooked_block', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
```

Comment:
Thinking to rename the filter to include the word "markup" in case we want to introduce a different filter that allows including a hooked block at an earlier stage (e.g. adding just the block type to the list of blocks returned from `get_hooked_blocks()`).

Maybe

```suggestion
				$markup .= apply_filters( 'insert_hooked_block_markup', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
```

or even

```suggestion
				$markup .= apply_filters( 'inject_hooked_block_markup', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
```

Reply:
29a42fb500